### PR TITLE
[test] Add new MultiDeviceOpenaccTest on Kesch and Tsa

### PR DESCRIPTION
**Pull request by lucamar**
The new check should address the request in CSCS RT #36652: in principle, the check could run on Kesch as well, however it makes use of recent `ReFrame` features not available in the module installed on the system (` reframe/2.19`).
An update of the existing check using `OpenAcc` should be possible instead of creating a new check, although one would need to take into account the target systems without multiple GPU devices per node.

### CODE REVIEW

File: cscs-checks/mch/set_openacc_device.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+# SPDX-License-Identifier: BSD-3-Clause
+
+import reframe as rfm
+import reframe.utility.sanity as sn
+
+
+@rfm.simple_test
+class SetOpenaccDevice(rfm.RegressionTest):
+    def __init__(self):
+        self.descr = 'Use OpenAcc, CUDA, MPI, and C++ on multi-device nodes'
```
Comment (vkarak) [inline cscs-checks/mch/set_openacc_device.py:13]: I think this description is not very accurate. Also should change the test name as well to something like `MultiDeviceOpenACCTest`?

File: cscs-checks/mch/set_openacc_device.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    def setflags(self):
+        if self.current_environ.name.startswith('PrgEnv-pgi'):
+            self.build_system.fflags += ['-acc']
+#            if self.current_system.name == 'kesch':
+#                self.build_system.fflags += ['-ta=tesla,cc35,cuda8.0']
+#                self.build_system.ldflags = [
+#                    '-acc', '-ta:tesla:cc35,cuda8.0', '-lstdc++',
+#                    '-L/global/opt/nvidia/cudatoolkit/8.0.61/lib64',
+#                    '-lcublas', '-lcudart'
+#                ]
```
Comment (vkarak) [inline cscs-checks/mch/set_openacc_device.py:58]: Do we need this?

File: cscs-checks/mch/set_openacc_device.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+class SetOpenaccDevice(rfm.RegressionTest):
+    def __init__(self):
+        self.descr = 'Use OpenAcc, CUDA, MPI, and C++ on multi-device nodes'
+        self.valid_systems = ['arolla:cn', 'tsa:cn']
+        self.valid_prog_environs = ['PrgEnv-cce', 'PrgEnv-pgi']
+        self.build_system = 'Make'
+        self.build_system.makefile = 'Makefile.set_openacc_device'
+        self.build_system.fflags = ['-O2']
+
+#        if self.current_system.name == 'kesch':
```
[the remark this answers is not in the document]
Comment (kraushm) [reply cscs-checks/mch/set_openacc_device.py:20]: I think, `kesch` could be removed completely.

File: cscs-checks/mch/set_openacc_device.py (hunk 2 above)
Comment (lucamar) [reply cscs-checks/mch/set_openacc_device.py:58]: The check should work on `Kesch` as well, in principle: however it failed for me, therefore I have commented these lines out to keep in mind the need to fix sooner or later.

File: cscs-checks/mch/set_openacc_device.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lucamar) [reply cscs-checks/mch/set_openacc_device.py:20]: @kraushm See my comments above

File: cscs-checks/mch/set_openacc_device.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+#            self.exclusive_access = True
+#            self.modules = ['cudatoolkit/8.0.61']
+#            self.num_tasks = 9
+#            self.num_tasks_per_node = 9
+#            self.num_gpus_per_node = 8
+#            self.build_system.options = ['NVCC_FLAGS="-arch=compute_37"']
+#            self.variables = {
+#                'MV2_USE_CUDA': '1',
+#                'G2G': '1'
+#            }
```
[the remark this answers is not in the document]
Comment (lucamar) [reply cscs-checks/mch/set_openacc_device.py:30]: See my comment below on `Kesch`: i'' try again now after implementing the requested changes. If I don't manage to make it work on `Kesch`, I can add a `FIXME` as suggested by @victorusu

**Comment by lucamar** [comment]
Do you know why `ReFrame` is failing today with the error below?
```
    import jsonschema
ModuleNotFoundError: No module named 'jsonschema'
```

**Comment by vkarak** [comment]
@lucamar You should either load one of the latest reframe modules or install reframe's requirements `pip install -r requirements.txt`.

**Comment by lucamar** [comment]
Thanks @vkarak, I have used `pip install -r requirements.txt --user` since I cannot write under the python folders installed by Jenkins and unfortunately on Tsa the only ReFrame module is `reframe/2.18`.

File: cscs-checks/mch/set_openacc_device.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lucamar) [reply cscs-checks/mch/set_openacc_device.py]: I have committed the changes that fix the execution on `Kesch` using both `PrgEnv-cray` and `PrgEnv-pgi`

File: cscs-checks/mch/set_openacc_device.py (hunk 2 above)
Comment (lucamar) [reply cscs-checks/mch/set_openacc_device.py]: I have committed the changes that fix the execution on `Kesch` using both `PrgEnv-cray` and `PrgEnv-pgi`

File: cscs-checks/mch/set_openacc_device.py (hunk 1 above)
Comment (lucamar) [reply cscs-checks/mch/set_openacc_device.py]: I have changed the name of the check as you suggested @vkarak and updated the description.

File: cscs-checks/mch/set_openacc_device.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucamar) [reply cscs-checks/mch/set_openacc_device.py]: I have committed the changes that fix the execution on `Kesch` using both `PrgEnv-cray` and `PrgEnv-pgi`.

File: cscs-checks/mch/multi_device_openacc.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+# SPDX-License-Identifier: BSD-3-Clause
+
+import reframe as rfm
+import reframe.utility.sanity as sn
+
+
+@rfm.simple_test
+class MultiDeviceOpenaccTest(rfm.RegressionTest):
+    def __init__(self):
+        self.descr = 'Allocate one accelerator per MPI task using OpenAcc on multi-device nodes with additional CUDA, MPI, and C++ calls'
```
Comment (vkarak) [inline cscs-checks/mch/multi_device_openacc.py:13]:
```suggestion
        self.descr = 'Allocate one accelerator per MPI task using OpenACC on multi-device nodes with additional CUDA, MPI, and C++ calls'
```

File: cscs-checks/mch/multi_device_openacc.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        self.valid_systems = ['arolla:cn', 'tsa:cn', 'kesch:cn']
+        self.valid_prog_environs = ['PrgEnv-cray', 'PrgEnv-pgi']
+        self.build_system = 'Make'
+        self.build_system.makefile = 'Makefile.multi_device_openacc'
+        self.build_system.fflags = ['-O2']
+
```
Comment (vkarak) [inline cscs-checks/mch/multi_device_openacc.py:19]: No need for this blank line.

File: cscs-checks/mch/multi_device_openacc.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            self.modules = ['cuda/10.1.243']
+            self.num_tasks = 9
+            self.num_tasks_per_node = 9
+            self.num_gpus_per_node = 8
+            self.build_system.options = ['NVCC_FLAGS="-arch=compute_70"']
+
+        self.executable = 'multi_device_openacc'
+        self.sanity_patterns = sn.assert_found(r'Test\sResult\s*:\s+OK',
+                                               self.stdout)
+        self.maintainers = ['LM']
```
Comment (vkarak) [inline cscs-checks/mch/multi_device_openacc.py:38]:
```suggestion
        self.maintainers = ['LM', 'AJ']
```

**Pull request merged**